#Data Wrangling on sales data

##data loading

In [158]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

df = pd.read_csv("train.csv")

In [159]:
df.head()

Row ID        Order ID  Order Date   Ship Date       Ship Mode Customer ID  \
0       1  CA-2017-152156  08/11/2017  11/11/2017    Second Class    CG-12520   
1       2  CA-2017-152156  08/11/2017  11/11/2017    Second Class    CG-12520   
2       3  CA-2017-138688  12/06/2017  16/06/2017    Second Class    DV-13045   
3       4  US-2016-108966  11/10/2016  18/10/2016  Standard Class    SO-20335   
4       5  US-2016-108966  11/10/2016  18/10/2016  Standard Class    SO-20335   

     Customer Name    Segment        Country             City       State  \
0      Claire Gute   Consumer  United States        Henderson    Kentucky   
1      Claire Gute   Consumer  United States        Henderson    Kentucky   
2  Darrin Van Huff  Corporate  United States      Los Angeles  California   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale     Florida   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale     Florida   

   Postal Code Region       Product ID         Category Sub-Category  \
0      42420.0  South  FUR-BO-10001798        Furniture    Bookcases   
1      42420.0  South  FUR-CH-10000454        Furniture       Chairs   
2      90036.0   West  OFF-LA-10000240  Office Supplies       Labels   
3      33311.0  South  FUR-TA-10000577        Furniture       Tables   
4      33311.0  South  OFF-ST-10000760  Office Supplies      Storage   

                                        Product Name     Sales  
0                  Bush Somerset Collection Bookcase  261.9600  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400  
2  Self-Adhesive Address Labels for Typewriters b...   14.6200  
3      Bretford CR4500 Series Slim Rectangular Table  957.5775  
4                     Eldon Fold 'N Roll Cart System   22.3680

In [160]:
df.shape

(9800, 18)

In [161]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9800 entries, 0 to 9799
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9800 non-null   int64  
 1   Order ID       9800 non-null   object 
 2   Order Date     9800 non-null   object 
 3   Ship Date      9800 non-null   object 
 4   Ship Mode      9800 non-null   object 
 5   Customer ID    9800 non-null   object 
 6   Customer Name  9800 non-null   object 
 7   Segment        9800 non-null   object 
 8   Country        9800 non-null   object 
 9   City           9800 non-null   object 
 10  State          9800 non-null   object 
 11  Postal Code    9789 non-null   float64
 12  Region         9800 non-null   object 
 13  Product ID     9800 non-null   object 
 14  Category       9800 non-null   object 
 15  Sub-Category   9800 non-null   object 
 16  Product Name   9800 non-null   object 
 17  Sales          9800 non-null   float64
dtypes: float

In [162]:
df.describe()

Row ID   Postal Code         Sales
count  9800.000000   9789.000000   9800.000000
mean   4900.500000  55273.322403    230.769059
std    2829.160653  32041.223413    626.651875
min       1.000000   1040.000000      0.444000
25%    2450.750000  23223.000000     17.248000
50%    4900.500000  58103.000000     54.490000
75%    7350.250000  90008.000000    210.605000
max    9800.000000  99301.000000  22638.480000

##Removing unnecessary columns

In [163]:
df.drop(["Row ID", "Order ID", "Customer ID", "Customer Name"], axis = 1, inplace=True)

In [164]:
df.head()

Order Date   Ship Date       Ship Mode    Segment        Country  \
0  08/11/2017  11/11/2017    Second Class   Consumer  United States   
1  08/11/2017  11/11/2017    Second Class   Consumer  United States   
2  12/06/2017  16/06/2017    Second Class  Corporate  United States   
3  11/10/2016  18/10/2016  Standard Class   Consumer  United States   
4  11/10/2016  18/10/2016  Standard Class   Consumer  United States   

              City       State  Postal Code Region       Product ID  \
0        Henderson    Kentucky      42420.0  South  FUR-BO-10001798   
1        Henderson    Kentucky      42420.0  South  FUR-CH-10000454   
2      Los Angeles  California      90036.0   West  OFF-LA-10000240   
3  Fort Lauderdale     Florida      33311.0  South  FUR-TA-10000577   
4  Fort Lauderdale     Florida      33311.0  South  OFF-ST-10000760   

          Category Sub-Category  \
0        Furniture    Bookcases   
1        Furniture       Chairs   
2  Office Supplies       Labels   
3        Furniture       Tables   
4  Office Supplies      Storage   

                                        Product Name     Sales  
0                  Bush Somerset Collection Bookcase  261.9600  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400  
2  Self-Adhesive Address Labels for Typewriters b...   14.6200  
3      Bretford CR4500 Series Slim Rectangular Table  957.5775  
4                     Eldon Fold 'N Roll Cart System   22.3680

##Checking Missing values

In [165]:
df.isnull().sum()

Order Date       0
Ship Date        0
Ship Mode        0
Segment          0
Country          0
City             0
State            0
Postal Code     11
Region           0
Product ID       0
Category         0
Sub-Category     0
Product Name     0
Sales            0
dtype: int64

##Handling missing values

In [166]:
df["Postal Code"].fillna(df.groupby("City")["Postal Code"].transform('first'),inplace=True)

<ipython-input-166-a3a131cd3f9e>:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df["Postal Code"].fillna(df.groupby("City")["Postal Code"].transform('first'),inplace=True)


In [167]:
df.isnull().sum()

Order Date      0
Ship Date       0
Ship Mode       0
Segment         0
Country         0
City            0
State           0
Postal Code     0
Region          0
Product ID      0
Category        0
Sub-Category    0
Product Name    0
Sales           0
dtype: int64

##Handling Duplicates

In [168]:
df =df.drop_duplicates()

In [169]:
df.shape

(9799, 14)

##Handling inappropriate data types

In [170]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9799 entries, 0 to 9799
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Order Date    9799 non-null   object 
 1   Ship Date     9799 non-null   object 
 2   Ship Mode     9799 non-null   object 
 3   Segment       9799 non-null   object 
 4   Country       9799 non-null   object 
 5   City          9799 non-null   object 
 6   State         9799 non-null   object 
 7   Postal Code   9799 non-null   float64
 8   Region        9799 non-null   object 
 9   Product ID    9799 non-null   object 
 10  Category      9799 non-null   object 
 11  Sub-Category  9799 non-null   object 
 12  Product Name  9799 non-null   object 
 13  Sales         9799 non-null   float64
dtypes: float64(2), object(12)
memory usage: 1.1+ MB


##Handling Dates

In [171]:
df["Ship Date"] = pd.to_datetime(df["Ship Date"], format="%d/%m/%Y", errors = 'coerce')

In [172]:
df["Order Date"] = pd.to_datetime(df["Order Date"], format="%d/%m/%Y", errors = 'coerce')

In [173]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9799 entries, 0 to 9799
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   Order Date    9799 non-null   datetime64[ns]
 1   Ship Date     9799 non-null   datetime64[ns]
 2   Ship Mode     9799 non-null   object        
 3   Segment       9799 non-null   object        
 4   Country       9799 non-null   object        
 5   City          9799 non-null   object        
 6   State         9799 non-null   object        
 7   Postal Code   9799 non-null   float64       
 8   Region        9799 non-null   object        
 9   Product ID    9799 non-null   object        
 10  Category      9799 non-null   object        
 11  Sub-Category  9799 non-null   object        
 12  Product Name  9799 non-null   object        
 13  Sales         9799 non-null   float64       
dtypes: datetime64[ns](2), float64(2), object(10)
memory usage: 1.1+ MB


##Handling categorical data

In [174]:
categorical_columns = [col for col in df.columns if df[col].dtypes == 'object']
categorical_columns

['Ship Mode',
 'Segment',
 'Country',
 'City',
 'State',
 'Region',
 'Product ID',
 'Category',
 'Sub-Category',
 'Product Name']

In [175]:

from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

for col in categorical_columns:
  df[col] = le.fit_transform(df[col])

In [176]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9799 entries, 0 to 9799
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   Order Date    9799 non-null   datetime64[ns]
 1   Ship Date     9799 non-null   datetime64[ns]
 2   Ship Mode     9799 non-null   int64         
 3   Segment       9799 non-null   int64         
 4   Country       9799 non-null   int64         
 5   City          9799 non-null   int64         
 6   State         9799 non-null   int64         
 7   Postal Code   9799 non-null   float64       
 8   Region        9799 non-null   int64         
 9   Product ID    9799 non-null   int64         
 10  Category      9799 non-null   int64         
 11  Sub-Category  9799 non-null   int64         
 12  Product Name  9799 non-null   int64         
 13  Sales         9799 non-null   float64       
dtypes: datetime64[ns](2), float64(2), int64(10)
memory usage: 1.1 MB


In [177]:
df.head()

Order Date  Ship Date  Ship Mode  Segment  Country  City  State  \
0 2017-11-08 2017-11-11          2        0        0   194     15   
1 2017-11-08 2017-11-11          2        0        0   194     15   
2 2017-06-12 2017-06-16          2        1        0   265      3   
3 2016-10-11 2016-10-18          3        0        0   153      8   
4 2016-10-11 2016-10-18          3        0        0   153      8   

   Postal Code  Region  Product ID  Category  Sub-Category  Product Name  \
0      42420.0       2          12         0             4           386   
1      42420.0       2          55         0             5           838   
2      90036.0       3         946         1            10          1432   
3      33311.0       2         319         0            16           366   
4      33311.0       2        1316         1            14           573   

      Sales  
0  261.9600  
1  731.9400  
2   14.6200  
3  957.5775  
4   22.3680

##Outlier Analysis for Sales

In [178]:
Q1 = df["Sales"].quantile(0.25)
Q3 = df["Sales"].quantile(0.75)

IQR = Q3 - Q1

lower_bound = Q1 - IQR * 1.5
upper_bound = Q3 + IQR * 1.5

df = df[(df["Sales"] > lower_bound) & (df["Sales"] < upper_bound)]

In [179]:
df.shape

(8654, 14)

##Feature Scaling

In [180]:
from sklearn.preprocessing import MinMaxScaler

scalar = MinMaxScaler()

df["Sales"]  = scalar.fit_transform(df[["Sales"]])

In [181]:
df.head()

Order Date  Ship Date  Ship Mode  Segment  Country  City  State  \
0 2017-11-08 2017-11-11          2        0        0   194     15   
2 2017-06-12 2017-06-16          2        1        0   265      3   
4 2016-10-11 2016-10-18          3        0        0   153      8   
5 2015-06-09 2015-06-14          3        0        0   265      3   
6 2015-06-09 2015-06-14          3        0        0   265      3   

   Postal Code  Region  Product ID  Category  Sub-Category  Product Name  \
0      42420.0       2          12         0             4           386   
2      90036.0       3         946         1            10          1432   
4      33311.0       2        1316         1            14           573   
5      90032.0       3         185         0             9           570   
6      90032.0       3         562         1             2          1128   

      Sales  
0  0.523245  
2  0.028364  
4  0.043866  
5  0.096872  
6  0.013678

##Validating sales by checking negative values of sales

In [182]:
df = df[df["Sales"] > 0]

In [183]:
df.shape

(8653, 14)

#Summary

1) Data loading and initial analysis

- check dataset using head() and print first five rows of data
- shape of dataset i.e. (9600,18)
- data loading using read_csv()
- information of dataset using info()
- statistical description of numerical features using describe()

2) Removing unnecessary features

- few features that was not necessary get droped using drop()
- droped features includes:

   - Row ID
   - Order ID
   - Customer ID
   - Customer Name

3) Missing Values

- checking missing values using isnull().sum() and found postal code has 11 missing entries
- missing entries of postal code get impute with the help of City as they are correlated with each other

4) Handling duplicates

- Checking duplicates and droping them using drop_duplicates()
- checking shape again so that we know about number of duplicate entries was present i.e. remaing entries was (9799, 14)

5) Convert columns in appropriate datatypes

- Ship Date and Order Date get converted into datetime from object
- All categorical features get converted into numerical values using labelencoder

6) Outlier ananlysis

- Sales had potential feature who had outliers
- Outlier analysis performed using IQR method
- Found that 1146 entries were outliers

7) Feature scaling

- Sales column get scale down in range of 0 to 1 using minmaxscalar

8) Data validation

- checking for entries from Sales which are less than 0
- and i found no such entries

